File: src/services/swarmService.js
/**
 * Swarm Service - Bounty and Consensus Management
 */

const Bounty = require('../models/Bounty');
const Asset = require('../models/Asset');
const logger = require('../utils/logger');
const { generateNodeId } = require('../utils/crypto');

const SWARM_CONFIG = {
  min_participants: 3,
  consensus_threshold: 0.67, // 2/3 majority
  decision_timeout: 24 * 60 * 60 * 1000, // 24 hours
  reputation_decay: 0.95 // Weekly decay
};

class SwarmService {
  /**
   * Create a new bounty task
   */
  async createBounty(creatorNodeId, bountyData) {
    const bountyId = `bounty_${Date.now()}_${Math.random().toString(36).substr(2, 8)}`;
    
    const bounty = new Bounty({
      bounty_id: bountyId,
      creator_node_id: creatorNodeId,
      title: bountyData.title,
      description: bountyData.description,
      task_type: bountyData.task_type,
      requirements: bountyData.requirements || {},
      rewards: bountyData.rewards,
      deadline: new Date(Date.now() + (bountyData.duration_hours || 24) * 3600000),
      tags: bountyData.tags || [],
      priority: bountyData.priority || 'medium'
    });

    await bounty.save();
    logger.info(`Swarm bounty created: ${bountyId} by ${creatorNodeId}`);
    
    return bounty;
  }

  /**
   * Join a bounty as participant
   */
  async joinBounty(bountyId, nodeId) {
    const bounty = await Bounty.findOne({ bounty_id: bountyId });
    if (!bounty) {
      throw new Error('Bounty not found');
    }

    if (bounty.status !== 'open') {
      throw new Error(`Cannot join bounty with status: ${bounty.status}`);
    }

    if (bounty.deadline < new Date()) {
      bounty.status = 'expired';
      await bounty.save();
      throw new Error('Bounty has expired');
    }

    // Check if already joined
    if (bounty.participants.some(p => p.node_id === nodeId)) {
      throw new Error('Already joined this bounty');
    }

    // Check max participants
    if (bounty.participants.length >= bounty.requirements.max_nodes) {
      throw new Error('Bounty is full');
    }

    bounty.participants.push({
      node_id: nodeId,
      joined_at: new Date()
    });

    // Update status if min participants reached
    if (bounty.participants.length >= SWARM_CONFIG.min_participants) {
      bounty.status = 'in_progress';
    }

    await bounty.save();
    logger.info(`Node ${nodeId} joined bounty ${bountyId}`);
    
    return bounty;
  }

  /**
   * Submit decision for bounty
   */
  async submitDecision(bountyId, nodeId, decision, reasoning, solutionAssetId = null) {
    const bounty = await Bounty.findOne({ bounty_id: bountyId });
    if (!bounty) {
      throw new Error('Bounty not found');
    }

    if (bounty.status !== 'in_progress') {
      throw new Error('Bounty is not in progress');
    }

    const participant = bounty.participants.find(p => p.node_id === nodeId);
    if (!participant) {
      throw new Error('Not a participant of this bounty');
    }

    if (participant.submitted_at) {
      throw new Error('Decision already submitted');
    }

    participant.decision = decision;
    participant.reasoning = reasoning;
    participant.submitted_at = new Date();

    // If solution provided, publish it
    if (solutionAssetId) {
      participant.solution_asset_id = solutionAssetId;
    }

    await bounty.save();
    logger.info(`Decision submitted for bounty ${bountyId} by ${nodeId}: ${decision}`);

    // Check if consensus can be reached
    await this.checkConsensus(bounty);
    
    return bounty;
  }

  /**
   * Check and process consensus
   */
  async checkConsensus(bounty) {
    const submitted = bounty.participants.filter(p => p.submitted_at);
    const total = bounty.participants.length;

    // Need at least min_participants decisions
    if (submitted.length < SWARM_CONFIG.min_participants) {
      return false;
    }

    // Count decisions
    const counts = { accept: 0, reject: 0, abstain: 0 };
    submitted.forEach(p => {
      counts[p.decision] = (counts[p.decision] || 0) + 1;
    });

    const maxDecision = Object.entries(counts).reduce((a, b) => a[1] > b[1] ? a : b);
    const ratio = maxDecision[1] / total;

    // Check if consensus reached
    if (ratio >= SWARM_CONFIG.consensus_threshold) {
      bounty.status = 'completed';
      bounty.completed_at = new Date();
      bounty.result = {
        consensus_reached: true,
        final_decision: maxDecision[0],
        winning_nodes: submitted
          .filter(p => p.decision === maxDecision[0])
          .map(p => p.node_id)
      };

      await bounty.save();
      await this.distributeRewards(bounty);
      
      logger.info(`Consensus reached for bounty ${bounty.bounty_id}: ${maxDecision[0]}`);
      return true;
    }

    // Check if all submitted but no consensus
    if (submitted.length === total) {
      bounty.status = 'completed';
      bounty.completed_at = new Date();
      bounty.result = {
        consensus_reached: false,
        final_decision: 'no_consensus',
        winning_nodes: []
      };

      await bounty.save();
      logger.info(`No consensus for bounty ${bounty.bounty_id}`);
      return false;
    }

    return false;
  }

  /**
   * Distribute rewards to winning nodes
   */
  async distributeRewards(bounty) {
    if (!bounty.result.consensus_reached) {
      return;
    }

    const { winning_nodes, final_decision } = bounty.result;
    
    if (final_decision === 'accept' && winning_nodes.length > 0) {
      const rewardPerNode = Math.floor(bounty.rewards.credits / winning_nodes.length);
      
      // TODO: Update node credits in database
      logger.info(`Distributed ${rewardPerNode} credits to ${winning_nodes.length} nodes for bounty ${bounty.bounty_id}`);
    }
  }

  /**
   * Get available bounties
   */
  async getAvailableBounties(filters = {}) {
    const query = {
      status: { $in: ['open', 'in_progress'] },
      deadline: { $gt: new Date() }
    };

    if (filters.task_type) {
      query.task_type = filters.task_type;
    }

    if (filters.tags) {
      query.tags = { $in: filters.tags };
    }

    if (filters.priority) {
      query.priority = filters.priority;
    }

    return await Bounty.find(query)
      .sort({ priority: -1, created_at: -1 })
      .limit(filters.limit || 50);
  }

  /**
   * Get bounty details
   */
  async getBounty(bountyId) {
    return await Bounty.findOne({ bounty_id: bountyId });
  }

  /**
   * Cancel bounty (only creator)
   */
  async cancelBounty(bountyId, nodeId) {
    const bounty = await Bounty.findOne({ bounty_id: bountyId });
    if (!bounty) {
      throw new Error('Bounty not found');
    }

    if (bounty.creator_node_id !== nodeId) {
      throw new Error('Only creator can cancel bounty');
    }

    if (bounty.status === 'completed') {
      throw new Error('Cannot cancel completed bounty');
    }

    bounty.status = 'cancelled';
    await bounty.save();
    
    logger.info(`Bounty ${bountyId} cancelled by ${nodeId}`);
    return bounty;
  }

  /**
   * Clean up expired bounties
   */
  async cleanupExpiredBounties() {
    const result = await Bounty.updateMany(
      {
        status: { $in: ['open', 'in_progress'] },
        deadline: { $lt: new Date() }
      },
      { $set: { status: 'expired' } }
    );

    if (result.modifiedCount > 0) {
      logger.info(`Marked ${result.modifiedCount} bounties as expired`);
    }

    return result.modifiedCount;
  }
}

module.exports = new SwarmService();

File: src/services/swarm_service.rs
use crate::models::bounty::{
    Bounty, BountyStatus, TaskType, Priority,
    CreateBountyRequest, CreateBountyResponse,
    JoinBountyRequest, JoinBountyResponse,
    SubmitDecisionRequest, SubmitDecisionResponse,
    ListBountiesResponse, BountySummary,
    GetBountyResponse, BountyDetail, ParticipantSummary,
    CancelBountyRequest, CancelBountyResponse,
};
use std::collections::HashMap;
use std::sync::{Arc, Mutex};

/// Swarm Bounty service
pub struct SwarmService {
    bounties: Arc<Mutex<HashMap<String, Bounty>>>,
}

impl SwarmService {
    pub fn new() -> Self {
        Self {
            bounties: Arc::new(Mutex::new(HashMap::new())),
        }
    }

    /// Create a new bounty
    pub fn create_bounty(
        &self,
        creator_node_id: String,
        req: CreateBountyRequest,
    ) -> Result<CreateBountyResponse, String> {
        let bounty_id = format!("bounty_{}_{}", 
            chrono::Utc::now().timestamp_millis(),
            uuid::Uuid::new_v4().to_string().split('-').next().unwrap_or("")
        );

        let mut bounty = Bounty::new(
            bounty_id.clone(),
            creator_node_id,
            req.title,
            req.description,
            req.task_type,
            req.duration_hours.unwrap_or(24),
        );

        // Apply optional fields
        if let Some(reqs) = req.requirements {
            bounty.requirements = reqs;
        }
        bounty.rewards = req.rewards;
        if let Some(tags) = req.tags {
            bounty.tags = tags;
        }
        if let Some(priority) = req.priority {
            bounty.priority = priority;
        }

        let mut bounties = self.bounties.lock().unwrap();
        bounties.insert(bounty_id.clone(), bounty);

        Ok(CreateBountyResponse {
            success: true,
            bounty_id,
            status: BountyStatus::Open,
        })
    }

    /// Join a bounty
    pub fn join_bounty(
        &self,
        node_id: String,
        req: JoinBountyRequest,
    ) -> Result<JoinBountyResponse, String> {
        let mut bounties = self.bounties.lock().unwrap();
        
        let bounty = bounties
            .get_mut(&req.bounty_id)
            .ok_or("Bounty not found")?;

        if !bounty.can_join(&node_id) {
            return Err("Cannot join bounty".to_string());
        }

        bounty.join(node_id);

        Ok(JoinBountyResponse {
            success: true,
            bounty_id: req.bounty_id,
            participants_count: bounty.participants.len(),
            status: bounty.status.clone(),
        })
    }

    /// Submit decision
    pub fn submit_decision(
        &self,
        node_id: String,
        req: SubmitDecisionRequest,
    ) -> Result<SubmitDecisionResponse, String> {
        let mut bounties = self.bounties.lock().unwrap();
        
        let bounty = bounties
            .get_mut(&req.bounty_id)
            .ok_or("Bounty not found")?;

        if bounty.status != BountyStatus::InProgress {
            return Err("Bounty is not in progress".to_string());
        }

        bounty.submit_decision(&node_id, req.decision, req.reasoning)?;

        Ok(SubmitDecisionResponse {
            success: true,
            bounty_id: req.bounty_id,
            consensus_reached: bounty.result.as_ref().map(|r| r.consensus_reached).unwrap_or(false),
            final_decision: bounty.result.as_ref().map(|r| r.final_decision.clone()),
            status: bounty.status.clone(),
        })
    }

    /// List available bounties
    pub fn list_bounties(
        &self,
        task_type: Option<TaskType>,
        tags: Option<Vec<String>>,
        priority: Option<Priority>,
        limit: i32,
    ) -> ListBountiesResponse {
        let bounties = self.bounties.lock().unwrap();
        
        let mut results: Vec<_> = bounties
            .values()
            .filter(|b| {
                matches!(b.status, BountyStatus::Open | BountyStatus::InProgress)
                    && b.deadline > chrono::Utc::now()
            })
            .filter(|b| {
                if let Some(ref tt) = task_type {
                    b.task_type == *tt
                } else {
                    true
                }
            })
            .filter(|b| {
                if let Some(ref t) = tags {
                    t.iter().any(|tag| b.tags.contains(tag))
                } else {
                    true
                }
            })
            .filter(|b| {
                if let Some(ref p) = priority {
                    b.priority == *p
                } else {
                    true
                }
            })
            .cloned()
            .collect();

        // Sort by priority and created_at
        results.sort_by(|a, b| {
            let priority_order = |p: &Priority| match p {
                Priority::Critical => 4,
                Priority::High => 3,
                Priority::Medium => 2,
                Priority::Low => 1,
            };
            priority_order(&b.priority)
                .cmp(&priority_order(&a.priority))
                .then_with(|| b.created_at.cmp(&a.created_at))
        });

        results.truncate(limit as usize);

        ListBountiesResponse {
            success: true,
            count: results.len(),
            bounties: results.into_iter().map(|b| BountySummary {
                bounty_id: b.bounty_id,
                title: b.title,
                description: truncate(&b.description, 200),
                task_type: b.task_type,
                status: b.status,
                rewards: b.rewards,
                deadline: b.deadline,
                participants_count: b.participants.len(),
                tags: b.tags,
                priority: b.priority,
            }).collect(),
        }
    }

    /// Get bounty details
    pub fn get_bounty(&self, bounty_id: &str) -> Result<GetBountyResponse, String> {
        let bounties = self.bounties.lock().unwrap();
        
        let bounty = bounties
            .get(bounty_id)
            .ok_or("Bounty not found")?;

        Ok(GetBountyResponse {
            success: true,
            bounty: BountyDetail {
                bounty_id: bounty.bounty_id.clone(),
                title: bounty.title.clone(),
                description: bounty.description.clone(),
                task_type: bounty.task_type.clone(),
                requirements: bounty.requirements.clone(),
                rewards: bounty.rewards.clone(),
                status: bounty.status.clone(),
                deadline: bounty.deadline,
                created_at: bounty.created_at,
                creator_node_id: bounty.creator_node_id.clone(),
                participants: bounty.participants.iter().map(|p| ParticipantSummary {
                    node_id: p.node_id.clone(),
                    joined_at: p.joined_at,
                    decision: p.decision.clone(),
                    submitted_at: p.submitted_at,
                }).collect(),
                result: bounty.result.clone(),
                tags: bounty.tags.clone(),
                priority: bounty.priority.clone(),
            },
        })
    }

    /// Cancel bounty
    pub fn cancel_bounty(
        &self,
        node_id: String,
        req: CancelBountyRequest,
    ) -> Result<CancelBountyResponse, String> {
        let mut bounties = self.bounties.lock().unwrap();
        
        let bounty = bounties
            .get_mut(&req.bounty_id)
            .ok_or("Bounty not found")?;

        bounty.cancel(&node_id)?;

        Ok(CancelBountyResponse {
            success: true,
            bounty_id: req.bounty_id,
            status: bounty.status.clone(),
        })
    }

    /// Cleanup expired bounties
    pub fn cleanup_expired(&self) -> usize {
        let mut bounties = self.bounties.lock().unwrap();
        let expired: Vec<_> = bounties
            .iter()
            .filter(|(_, b)| {
                matches!(b.status, BountyStatus::Open | BountyStatus::InProgress)
                    && b.deadline < chrono::Utc::now()
            })
            .map(|(id, _)| id.clone())
            .collect();

        let count = expired.len();
        for id in expired {
            if let Some(bounty) = bounties.get_mut(&id) {
                bounty.status = BountyStatus::Expired;
            }
        }

        count
    }
}

impl Default for SwarmService {
    fn default() -> Self {
        Self::new()
    }
}

fn truncate(s: &str, max_len: usize) -> String {
    if s.len() <= max_len {
        s.to_string()
    } else {
        format!("{}...", &s[..max_len])
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::models::bounty::Rewards;

    fn create_test_bounty_req() -> CreateBountyRequest {
        CreateBountyRequest {
            title: "Test Bounty".to_string(),
            description: "Test description".to_string(),
            task_type: TaskType::CodeReview,
            requirements: None,
            rewards: Rewards { credits: 100, reputation: 0.05 },
            duration_hours: Some(24),
            tags: None,
            priority: None,
        }
    }

    #[test]
    fn test_create_bounty() {
        let service = SwarmService::new();
        let req = create_test_bounty_req();
        
        let result = service.create_bounty("node_1".to_string(), req);
        assert!(result.is_ok());
        
        let response = result.unwrap();
        assert!(response.success);
        assert!(!response.bounty_id.is_empty());
    }

    #[test]
    fn test_join_bounty() {
        let service = SwarmService::new();
        
        // Create bounty
        let create_result = service.create_bounty("node_1".to_string(), create_test_bounty_req());
        let bounty_id = create_result.unwrap().bounty_id;
        
        // Join with different node
        let join_req = JoinBountyRequest { bounty_id: bounty_id.clone() };
        let join_result = service.join_bounty("node_2".to_string(), join_req);
        
        assert!(join_result.is_ok());
        assert_eq!(join_result.unwrap().participants_count, 1);
    }

    #[test]
    fn test_cancel_bounty() {
        let service = SwarmService::new();
        
        let create_result = service.create_bounty("node_1".to_string(), create_test_bounty_req());
        let bounty_id = create_result.unwrap().bounty_id;
        
        let cancel_req = CancelBountyRequest { bounty_id };
        let cancel_result = service.cancel_bounty("node_1".to_string(), cancel_req);
        
        assert!(cancel_result.is_ok());
        assert!(matches!(cancel_result.unwrap().status, BountyStatus::Cancelled));
    }
}
